File: Gregory Coefficients/Series 3/series_3.py
from decimal import Decimal
from time import perf_counter

n: int = 0
approximationIntermediate: float = 0
approximation: Decimal = 0
previous: Decimal = 0
finalAccuracy: int = 0
totalComputationTime: float = 0
Coeffecients: list[float] = []
iterationStartTime: float = 0
iterationEndTime: float = 0

def GregoryCoeffecient(num: int) -> float:
    if num == 1 : return 1/2

    result: float = 0 - sum(Coeffecients[k + 1] / (num - k + 1) for k in range(1, num))
    if result + (1 / (num + 1)) not in Coeffecients: Coeffecients += result + (1/ (num + 1))
    return result + (1 / (num + 1))

while True:

    iterationStartTime = perf_counter()
    approximationIntermediate += pow(-1, n) * (GregoryCoeffecient(3*n + 3) + GregoryCoeffecient(3*n + 4))
    approximation = pow(3, 1/2) / (1/2 - approximationIntermediate)
    iterationEndTime = perf_counter()

    print(f"\nIteration {n + 1}")
    print(f"Approximation = {approximation}")

    for i, char in enumerate("3.141592653589793238462643383279502884197169399375"): # using str(pi) instead of writing the whole string like I have done here somehow displays a different number??? idk
        if char != str(approximation)[i]:
            if i < 2: print("No accurate decimal places")
            else:
                print(f"{i - 2} correct decimal place(s)")
                finalAccuracy = i - 2
            break

    print(f"Iteration duration: {iterationEndTime - iterationStartTime}  seconds")
    totalComputationTime += iterationEndTime - iterationStartTime

    deviation: Decimal = approximation - previous
    if deviation == 0:
        print("Negligible deviation (terminating the program)\n")
        break
    elif deviation != approximation: print(f"Deviation from previous iteration: {deviation}") # this if-statement prevents a deviation from being shown during the first iteration.

    previous = approximation
    n += 1

print(f"\n\nComputed {finalAccuracy} correct decimal places in {totalComputationTime} seconds and {n + 1} iterations.\n")

File: Factorials/Rising Factorial/Series 3/rs/series_3.rs
use std::time::{Duration, Instant};

fn main() {

    let mut iterator: u16 = 0;
    let mut approximation_inverted: f64 = 0.0;
    let mut approximation: f64;
    let mut previous_approximation: f64 = 0.0;
    let mut deviation: f64;
    let mut final_accuracy: u8 = 0;
    let mut start_time: Instant;
    let mut iteration_time: Duration;
    let mut total_computation_time: Duration = Duration::from_secs(0); // Duration::from_secs(0) is just a temporary value in order to initialize the total_computation_time variable.

    loop {

        start_time = Instant::now();
        approximation_inverted += (f64::powf(2.0/27.0, iterator as f64) * (15*iterator + 2) as f64 * rising_factorial(1.0/2.0, iterator) * rising_factorial(1.0/3.0, iterator) * rising_factorial(2.0/3.0, iterator)) / u128::pow(factorial(iterator), 3) as f64;
        approximation = 27.0 / 4.0 / approximation_inverted;
        iteration_time = start_time.elapsed(); // only the mathematical computations are considered in the total computation time, and everything else like calculating the deviation and accuracy is not considered.

        println!("Iteration {}", iterator + 1);
        println!("Approximation = {:.51}", approximation);

        let mut i: u8 = 0;
        for c in String::from("3.141592653589793238462643383279502884197169399375105").chars() {
            if c != match approximation.to_string().chars().nth(i as usize) {
                Some(c2) => c2,
                None => char::from(0)
            } {
                if i < 2 { println!("No accurate decimal places.") }
                else {
                    println!("{} correct decimal place(s)", i - 2);
                    final_accuracy = i - 2;
                }
                break;
            }

            i += 1;
        }

        deviation = approximation - previous_approximation;
        if deviation.abs() < 1e-50 {
            println!("\nSummation converged. Terminating program...");
            break;
        }
        else { println!("Deviation from previous iteration: {:.51}\n", deviation) }

        println!("Iteration duration: {:?}\n", iteration_time);
        total_computation_time += iteration_time;

        previous_approximation = approximation;
        iterator += 1;

    }

    println!("Computed {} correct decimal places in {:?} and {} iterations.\n", final_accuracy, total_computation_time, iterator + 1);
}

fn factorial(num: u16) -> u128 {
    if num == 0 { return 1; }
    return num as u128 * factorial(num - 1);
}

fn rising_factorial(x: f64, n: u16) -> f64 {
    let mut result: f64 = 1.0;

    if n == 0 { return 1.0; }

    for i in 0..n { result *= x + i as f64 }
    return result;
}